### ISSUE TITLE
Incorrect regex for floating point numbers

### ISSUE THREAD

**Issue by diesirae68** [php/doc-en#2711]
From manual page: https://www.php.net/manual/en/language.types.float.php
Hi! I think the regex for floating point numbers is wrong.

Please, replace

`DNUM          ([0-9]*(_[0-9]+)*[\.]{LNUM}) | ({LNUM}[\.][0-9]*(_[0-9]+)*)`

with

`DNUM          (([0-9]+(_[0-9]+)*)?\.{LNUM}) | ({LNUM}\.([0-9]+(_[0-9]+)*)?)`

The regex specified in the documentation accepts such numbers as correct, but they are incorrect:
_123.456
123._456

And need to fix this article in all languages. Thanx!

**Comment by nathansalter** [comment]
I actually think the REGEX can be simplified further to this:

```
LNUM = [0-9](_?[0-9]+)*
DNUM = [+-]?{LNUM}(\.{LNUM})?
EXPONENT_DNUM = {DNUM}[Ee]{DNUM}
```

Which might make it a bit easier to read

**Comment by Girgias** [comment]
> I actually think the REGEX can be simplified further to this:
> 
> ```
> LNUM = [0-9](_?[0-9]+)*
> DNUM = [+-]?{LNUM}(\.{LNUM})?
> EXPONENT_DNUM = {DNUM}[Ee]{DNUM}
> ```
> 
> Which might make it a bit easier to read

This doesn't seem to accept ``-.5`` as a valid float.

**Comment by diesirae68** [comment]
> > I actually think the REGEX can be simplified further to this:
> > ```
> > LNUM = [0-9](_?[0-9]+)*
> > DNUM = [+-]?{LNUM}(\.{LNUM})?
> > EXPONENT_DNUM = {DNUM}[Ee]{DNUM}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >               	
> > Which might make it a bit easier to read
> 
> This doesn't seem to accept `-.5` as a valid float.

You're right. But `[+-]?(({LNUM})?\.{LNUM})|({LNUM}\.({LNUM})?)` works correctly (this is a shorter expression than I originally wrote).

You can check this code:

```
$LNUM = '[0-9]+(_[0-9]+)*';
$DNUM = "[+-]?(({$LNUM})?\.{$LNUM})|({$LNUM}\.({$LNUM})?)";
var_dump(preg_match("/^{$DNUM}$/", '-.5'));
```

**Comment by Girgias** [comment]
This looks correct, can you make a PR?

**Comment by thegallagher** [comment]
Sorry to jump in here at the last minute, but I'd like to offer an alternative opinion on this.

I believe the purpose of this section of the documentation is to describe the structure of a PHP float literal. The description doesn't need to be useful for other purposes (eg. input validation) and it doesn't need to be a valid PCRE regular expression.

The [grammar](https://github.com/php/php-src/blob/master/Zend/zend_language_scanner.l#L1359) used by the PHP lexer is:

```
LNUM            [0-9]+(_[0-9]+)*
DNUM            ({LNUM}?"."{LNUM})|({LNUM}"."{LNUM}?)
EXPONENT_DNUM   (({LNUM}|{DNUM})[eE][+-]?{LNUM})
```

The above grammar or something derived from it should be used to correct the documentation.

While obviously correct, it is unclear:

  - that both `DNUM` and `EXPONENT_DNUM` define complete float literals
  - that `LNUM` defines only part of a float literal
  - what `LNUM` and `DNUM` are

To clarify the above grammar and for consistency with the [integer documentation](https://www.php.net/manual/en/language.types.integer.php), I offer the following alternative:

```
digits     : [0-9]+(_[0-9]+)*

decimal    : (digits? "." digits)
           | (digits "." digits?)

scientific : (digits | decimal) [eE][+-]? digits

float      : decimal
           | scientific
```

For me, the above grammar is easier to read and therefore better achieves the purpose I've described for the documentation, but I'd like to hear other thoughts.

**Comment by diesirae68** [comment]
> Sorry to jump in here at the last minute, but I'd like to offer an alternative opinion on this.
> 
> I believe the purpose of this section of the documentation is to describe the structure of a PHP float literal. The description doesn't need to be useful for other purposes (eg. input validation) and it doesn't need to be a valid PCRE regular expression.
> 
> The [grammar](https://github.com/php/php-src/blob/master/Zend/zend_language_scanner.l#L1359) used by the PHP lexer is:
> 
> ```
> LNUM            [0-9]+(_[0-9]+)*
> DNUM            ({LNUM}?"."{LNUM})|({LNUM}"."{LNUM}?)
> EXPONENT_DNUM   (({LNUM}|{DNUM})[eE][+-]?{LNUM})
> ```
> 
> The above grammar or something derived from it should be used to correct the documentation.
> 
> While obviously correct, it is unclear:
> 
> * that both `DNUM` and `EXPONENT_DNUM` define complete float literals
> * that `LNUM` defines only part of a float literal
> * what `LNUM` and `DNUM` are
> 
> To clarify the above grammar and for consistency with the [integer documentation](https://www.php.net/manual/en/language.types.integer.php), I offer the following alternative:
> 
> ```
> digits     : [0-9]+(_[0-9]+)*
> 
> decimal    : (digits? "." digits)
>            | (digits "." digits?)
> 
> scientific : (digits | decimal) [eE][+-]? digits
> 
> float      : decimal
>            | scientific
> ```
> 
> For me, the above grammar is easier to read and therefore better achieves the purpose I've described for the documentation, but I'd like to hear other thoughts.

Your description of the floating point number format would be ideal for documentation. I would choose this description.

**Comment by Girgias** [comment]
People, if you have suggestions, you can just send a PR.

**Comment by diesirae68** [comment]
> People, if you have suggestions, you can just send a PR.

Done! :)

**Comment by diesirae68** [comment]
Do I need to submit a PR for documentation in other languages?

**Comment by Girgias** [comment]
> Do I need to submit a PR for documentation in other languages?

You _can_, but this change is tracked via our RevCheck for all the languages via: https://doc.php.net

**Comment by Girgias** [comment]
Fixed via #2711
